registry/storage/garbagecollect.go

```diff
@@ -47,6 +61,21 @@ func MarkAndSweep(ctx context.Context, storageDriver driver.StorageDriver, regis
 		}
 
 		err = manifestEnumerator.Enumerate(ctx, func(dgst digest.Digest) error {
+			if opts.RemoveUntagged {
+				tags, err := repository.Tags(ctx).Lookup(ctx, distribution.Descriptor{Digest: dgst})
+				if err != nil {
+					return fmt.Errorf("failed to retrieve tags for digest %v: %v", dgst, err)
+				}
+				if len(tags) == 0 {
+					emit("manifest eligible for deletion: %s", dgst)
+					allTags, err := repository.Tags(ctx).All(ctx)
```

Comment:
Why does it look up tags again here? This may be worthy of a comment

Reply:
row 65 fetches the information "retrieve tags where this manifest is tagged currently". And if we want delete manifest it should return 0 tags in it (row 69). After that this allTags is fetching the information where this manifest has been tagged, but its not anymore maybe the latest one.

Example:

```
% docker tag alpine:3.4 localhost:5000/alpine:latest
% docker push localhost:5000/alpine:latest
The push refers to a repository [localhost:5000/alpine]
8539d1fe4fab: Pushed
latest: digest: sha256:0c8da65d4500db7421cc95021208989b850804c9214352f48dc5ffbd351387f4 size: 528
% docker tag alpine:latest localhost:5000/alpine:latest
% docker push localhost:5000/alpine:latest
The push refers to a repository [localhost:5000/alpine]
3fb66f713c9f: Pushed
latest: digest: sha256:5cb04fce748f576d7b72a37850641de8bd725365519673c643ef2d14819b42c6 size: 528
% docker exec 86e1bb8f6bba ls /var/lib/registry/docker/registry/v2/repositories/alpine/_manifests/tags/latest/index/sha256
0c8da65d4500db7421cc95021208989b850804c9214352f48dc5ffbd351387f4
5cb04fce748f576d7b72a37850641de8bd725365519673c643ef2d14819b42c6
```

In this case we could get rid off manifest 0c8da65d4500db7421cc95021208989b850804c9214352f48dc5ffbd351387f4 because it is not currently anymore "latest" tag in any of the tags. However, when we retrieve allTags for that manifest it should return latest, because it is there in "history".

I added debug print for that manifest:
len tags %!s(int=0) len alltags %!s(int=1) tags [] alltags [latest]

Now if I run -m for that registry it will print row:

time=“2018-01-04T07:38:19.115077762Z” level=info msg=“deleting manifest tag reference: /docker/registry/v2/repositories/alpine/_manifests/tags/latest/index/sha256/0c8da65d4500db7421cc95021208989b850804c9214352f48dc5ffbd351387f4" environment=development go.version=go1.8.5 instance.id=c60a26f1-5a00-4b1f-a92e-e56ddc27295e service=registry

So answer for your question: yes this allTags is needed as well, otherwise _manifest/tags/[tag]/index/sha256/ folder is not cleaned correctly